File: app/src/main/java/com/nityam/movsync/data/sync/SyncEngine.kt
package com.nityam.movsync.data.sync

import androidx.media3.common.Player
import com.google.firebase.database.ValueEventListener
import com.nityam.movsync.data.firebase.FirebaseSync
import com.nityam.movsync.data.model.HeartbeatState
import com.nityam.movsync.data.model.SyncState
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch

class SyncEngine(
    private val firebaseSync: FirebaseSync,
    private val driftCorrector: DriftCorrector
) {
    private var heartbeatJob: Job? = null
    private var rttJob: Job? = null
    private var syncListener: ValueEventListener? = null
    private var heartbeatListener: ValueEventListener? = null
    private var playerListener: Player.Listener? = null
    private var applyingRemoteCommand = false

    fun start(
        roomCode: String,
        userId: String,
        isHost: Boolean,
        player: Player,
        scope: CoroutineScope,
        allowControlsFlow: kotlinx.coroutines.flow.StateFlow<Boolean>,
        onStatus: (SyncStatus) -> Unit
    ) {
        stop(roomCode, player)
        
        // 1. Listen to Sync Queue (Both Host & Participant)
        syncListener = firebaseSync.listenToSync(roomCode) { state ->
            if (state.senderId == userId) return@listenToSync
            applyingRemoteCommand = true
            driftCorrector.resetOnCommand()
            when (state.command) {
                "play" -> {
                    player.seekTo(state.position)
                    player.play()
                }
                "pause" -> {
                    player.seekTo(state.position)
                    player.pause()
                }
                "seek" -> player.seekTo(state.position)
            }
            player.setPlaybackSpeed(state.speed)
            applyingRemoteCommand = false
        }

        // 2. Explicit Broadcast function added instead of Player.Listener

        // 3. Heartbeat & Drift Correction
        if (isHost) {
            heartbeatJob = scope.launch {
                while (true) {
                    firebaseSync.writeHeartbeat(
                        roomCode,
                        HeartbeatState(
                            position = player.currentPosition,
                            isPlaying = player.isPlaying
                        )
                    )
                    delay(500L)
                }
            }
        } else {
            // Start periodic RTT measurement
            rttJob = scope.launch {
                while (true) {
                    firebaseSync.measureRtt(roomCode)
                    delay(30_000L)
                }
            }
            // Initial RTT measurement
            scope.launch { firebaseSync.measureRtt(roomCode) }

            heartbeatListener = firebaseSync.listenToHeartbeat(roomCode) { heartbeat ->
                if (applyingRemoteCommand) return@listenToHeartbeat

                val oneWayLatency = firebaseSync.getOneWayLatency()
                val adjustedPosition = heartbeat.position
                    + (firebaseSync.getEstimatedServerTime() - heartbeat.timestamp)
                    - oneWayLatency

                if (heartbeat.isPlaying && !player.isPlaying) {
                    player.play()
                } else if (!heartbeat.isPlaying && player.isPlaying) {
                    player.pause()
                }
                driftCorrector.apply(player, adjustedPosition.coerceAtLeast(0L), !player.isPlaying, onStatus)
            }
            scope.launch {
                firebaseSync.getHeartbeatOnce(roomCode)?.let { heartbeat ->
                    val oneWayLatency = firebaseSync.getOneWayLatency()
                    val adjustedPosition = heartbeat.position
                        + (firebaseSync.getEstimatedServerTime() - heartbeat.timestamp)
                        - oneWayLatency
                    player.seekTo(adjustedPosition.coerceAtLeast(0L))
                    if (heartbeat.isPlaying) player.play() else player.pause()
                }
            }
        }
    }

    fun stop(roomCode: String, player: Player? = null) {
        heartbeatJob?.cancel()
        heartbeatJob = null
        rttJob?.cancel()
        rttJob = null
        syncListener?.let { firebaseSync.removeSyncListener(roomCode, it) }
        heartbeatListener?.let { firebaseSync.removeHeartbeatListener(roomCode, it) }
        syncListener = null
        heartbeatListener = null
        playerListener?.let { player?.removeListener(it) }
        playerListener = null
        applyingRemoteCommand = false
    }

    suspend fun broadcastCommand(
        roomCode: String,
        userId: String,
        command: String,
        position: Long,
        speed: Float
    ) {
        firebaseSync.writeSyncCommand(
            roomCode,
            SyncState(
                command = command,
                position = position,
                speed = speed,
                senderId = userId
            )
        )
    }
}

File: web/src/sync-engine.js
/**
 * SyncEngine — Port of SyncEngine.kt
 * Manages playback sync between host and participants via Firebase RTDB.
 */

import * as firebaseSync from './firebase-sync.js';
import { cleanup as cleanupDrift, resetDriftOnCommand } from './drift-corrector.js';

let heartbeatInterval = null;
let rttInterval = null;
let unsubSync = null;
let unsubHeartbeat = null;
let applyingRemoteCommand = false;

/**
 * Start the sync engine for a room.
 * @param {Object} opts
 * @param {string} opts.roomCode
 * @param {string} opts.userId
 * @param {boolean} opts.isHost
 * @param {HTMLVideoElement} opts.video
 * @param {Function} opts.onStatus - callback(status: 'synced'|'correcting'|'reconnecting')
 */
export function start({ roomCode, userId, isHost, video, onStatus }) {
  stop(roomCode);
  firebaseSync.getServerTime(); // Ensure server time offset starts observing

  // 1. Listen to sync commands (both host & participant)
  unsubSync = firebaseSync.listenToSync(roomCode, (state) => {
    if (state.senderId === userId) return;
    applyingRemoteCommand = true;
    resetDriftOnCommand();
    switch (state.command) {
      case 'play':
        video.currentTime = state.position / 1000;
        video.play().catch(() => {});
        break;
      case 'pause':
        video.currentTime = state.position / 1000;
        video.pause();
        break;
      case 'seek':
        video.currentTime = state.position / 1000;
        break;
    }
    if (state.speed && state.speed > 0) {
      video.playbackRate = state.speed;
    }
    applyingRemoteCommand = false;
  });

  // 2. Heartbeat & Drift Correction
  if (isHost) {
    // Host writes heartbeat every 500ms
    heartbeatInterval = setInterval(() => {
      firebaseSync.writeHeartbeat(roomCode, {
        position: Math.round(video.currentTime * 1000),
        isPlaying: !video.paused,
      });
    }, 500);
  } else {
    // Initial RTT measurement
    firebaseSync.measureRtt(roomCode, userId);
    // Periodic RTT re-measurement
    rttInterval = setInterval(() => {
      firebaseSync.measureRtt(roomCode, userId);
    }, 30000);

    // Participant listens to heartbeat and applies drift correction
    unsubHeartbeat = firebaseSync.listenToHeartbeat(roomCode, (heartbeat) => {
      if (applyingRemoteCommand) return;

      const oneWayLatency = firebaseSync.getOneWayLatency();
      const serverNow = firebaseSync.getServerTime();
      const adjustedPosition = heartbeat.position + (serverNow - heartbeat.timestamp) + oneWayLatency;

      if (heartbeat.isPlaying && video.paused) {
        video.play().catch(() => {});
      } else if (!heartbeat.isPlaying && !video.paused) {
        video.pause();
      }

      applyDriftCorrection(video, adjustedPosition, !heartbeat.isPlaying, onStatus);
    });

    // Late joiner: fetch heartbeat once and catch up
    firebaseSync.getHeartbeatOnce(roomCode).then((heartbeat) => {
      if (heartbeat) {
        const oneWayLatency = firebaseSync.getOneWayLatency();
        const adjustedPosition = heartbeat.position + (firebaseSync.getServerTime() - heartbeat.timestamp) + oneWayLatency;
        video.currentTime = Math.max(adjustedPosition / 1000, 0);
        if (heartbeat.isPlaying) {
          video.play().catch(() => {});
        } else {
          video.pause();
        }
      }
    });
  }
}

/**
 * Stop the sync engine and clean up listeners.
 */
export function stop(roomCode) {
  if (heartbeatInterval) {
    clearInterval(heartbeatInterval);
    heartbeatInterval = null;
  }
  if (rttInterval) {
    clearInterval(rttInterval);
    rttInterval = null;
  }
  if (unsubSync) {
    unsubSync();
    unsubSync = null;
  }
  if (unsubHeartbeat) {
    unsubHeartbeat();
    unsubHeartbeat = null;
  }
  cleanupDrift();
  applyingRemoteCommand = false;
}

/**
 * Broadcast a sync command to the room.
 */
export async function broadcastCommand(roomCode, userId, command, positionMs, speed = 1) {
  await firebaseSync.writeSyncCommand(roomCode, {
    command,
    position: positionMs,
    speed,
    senderId: userId,
  });
}

/**
 * Check if a remote command is currently being applied (to avoid echo).
 */
export function isApplyingRemote() {
  return applyingRemoteCommand;
}
